kafka-cron/producer/main.go

```diff
@@ -0,0 +1,192 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/signal"
+	"syscall"
+
+	kingpin "gopkg.in/alecthomas/kingpin.v2"
+
+	"github.com/Shopify/sarama"
+	"github.com/berkeli/kafka-cron/types"
+	"github.com/goccy/go-yaml"
+	"github.com/google/uuid"
+	"github.com/robfig/cron/v3"
+)
+
+var (
+	brokerList  = kingpin.Flag("brokerList", "List of brokers to connect").Default("localhost:9092").Strings()
+	topicPrefix = kingpin.Flag("topicPrefix", "Topic prefix, e.g. jobs will create topics for each cluster in the format jobs-cluster-a").Default("jobs").String()
+	cronPath    = kingpin.Flag("cronPath", "Path to cron file").Default("./cron.yaml").String()
+	partitions  = kingpin.Flag("partitions", "Number of partitions").Default("1").Int32()
+)
+
+func main() {
+	InitMonitoring(2112)
+
+	kingpin.Parse()
+	config := sarama.NewConfig()
+
+	admin, err := sarama.NewClusterAdmin(*brokerList, config)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	config.Producer.RequiredAcks = sarama.WaitForAll
+	config.Producer.Return.Successes = true
+	producer, err := sarama.NewSyncProducer(*brokerList, config)
+	if err != nil {
+		log.Panic(err)
+	}
+	defer func() {
+		if err := producer.Close(); err != nil {
+			log.Panic(err)
+		}
+		if err := admin.Close(); err != nil {
+			log.Panic(err)
+		}
+	}()
+	cmds, err := ReadConfig(*cronPath)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	err = CreateTopics(admin, cmds)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	err = ScheduleJobs(producer, cmds)
+
+	if err != nil {
+		log.Panic(err)
+	}
+
+	sigs := make(chan os.Signal, 1)
+
+	signal.Notify(sigs, syscall.SIGINT, syscall.SIGTERM)
+
+	<-sigs
+}
+
+func CreateTopics(admin sarama.ClusterAdmin, cmds []types.Command) error {
+	// Create topics for all unique clusters
+	clusters := make([]string, 0)
+
+	for _, cmd := range cmds {
+		for _, cluster := range cmd.Clusters {
+			if !contains(clusters, cluster) {
```

Comment:
I'd typically expect a `map[string]struct{}` rather than a `[]string` for this kind of check - can you talk through why?

Reply:
Has been updated to use a map now: 

1. using a map is faster, we don't need to interate over all elements. In terms of big O it will be O(1) for map and O(n) for list (only for contains function). Taking into consideration it's also used inside 2 loops, it could be quite slow.
2. It opens doors for additional checks, such as max_retries_allowed per cluster.